src/bioclip_vector_db/vector_db.py

```diff
@@ -0,0 +1,187 @@
+"""
+Author: Sreejith Menon 
+Executable script for setting up a database of vectors from the bioclip dataset
+"""
+import argparse
+import chromadb
+import datasets
+import enum
+import logging
+import os
+import torch 
+
+from bioclip.predict import TreeOfLifeClassifier
+from tqdm import tqdm
+from typing import List
+
+_DEFAULT_OUTPUT_DIR = os.path.join(os.getcwd(), 
+                                   "vector_db")
+_LOG_FORMAT = "[%(asctime)s] [%(levelname)s] [%(name)s] %(message)s"
+logging.basicConfig(level=logging.INFO, format=_LOG_FORMAT)
+logger = logging.getLogger()
+
+def _get_device() -> torch.device:
+    if torch.cuda.is_available():
+        logger.info("CUDA is available")
+        return torch.device("cuda")
+    elif torch.mps.is_available():
+        logger.info("MPS is available")
+        return torch.device("mps")
+    else:
+        logger.warning("CUDA and MPS are not available. Default to CPU")
+        return torch.device("cpu")
+    
+class HfDatasetType(enum.Enum):
+    BIRD = "Somnath01/Birds_Species"
+    TREE_OF_LIFE = "imageomics/TreeOfLife-10M"
+
+class BioclipVectorDatabase: 
+    def __init__(self, dataset_type: HfDatasetType, 
+                 collection_dir: str, 
+                 split: str):
+        self._dataset_type = dataset_type
+        self._classifier = TreeOfLifeClassifier(
+            device=_get_device())
+        self._dataset = None
+        self._collection_dir = collection_dir
+        self._client = None
+        self._collection = None
+        
+        self._prepare_dataset(
+            split=split)
+        self._init_collection()
+        
+    def _prepare_dataset(self, 
+                         split: str) -> datasets.Dataset:
+        """ Loads the dataset from Hugging Face to memory. """
+        if split is None:
+            raise ValueError("Split cannot be None. Please provide a valid split.")
+        
+        logger.info(f"Loading dataset: {self._dataset_type.value} for split: {split}")
+        self._dataset = datasets.load_dataset(self._dataset_type.value, 
+                                     split=split, 
+                                     streaming=False)
+        logger.info(f"Dataset loaded with {len(self._dataset)} records.")
+    
+    def _init_collection(self):
+        """ Initializes the collection for storing the vectors. """
+        self._client = chromadb.PersistentClient(path=self._collection_dir)
+
+        self._collection = self._client.get_or_create_collection(
+            name=self._dataset_type.name, 
+            metadata={
+                "hnsw:space": "cosine",
```

Comment:
I believe this should be inner product to match training.

Reply:
done